### PR TITLE
[PAPP-32566] Censys: Fix/Refactor - Integration of new API v2 search for ips and certs actions

### PR DESCRIPTION
**Pull request by mposluszny-splunk**
Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [ ] Testing of all the changes has been performed (for bug fixes / features)
- [ ] The manual_readme_content.md has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [ ] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [x] Bugfix
- [ ] Feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe):

## Security Considerations (REQUIRED)
- [x] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers),
  please document them in the `manual_readme_content.md`.
- [x] If this is a new connector or you are adding new actions
    - [x] Please document in the `manual_readme_content.md` all methods (eg, OAuth) used to authenticate
      with the service that the connector is integrating with.
    - [x] If any actions are unable to run on SOAR Cloud, please document this in the `manual_readme_content.md`.
- [x] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [x] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [x] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.

## What is the current behavior? (OPTIONAL)
- query_certificate is not working due to old implemenation, query_ip works api v2 changes were implemented

## What is the new behavior? (OPTIONAL)
- both query_certificate and query_ip are using the same module to handle pagination and delivery of results
- helper methods were moved to censys_validation.py package to make main code more clear
- censys api v2 search module is located under censys_search.py

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

File: censys_search.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        cursor_q = "" if cursor == "" else f"&cursor={cursor}"
+        endpoint_url = self._endpoint.format(
+            dataset=dataset, q=q, per_page=min(limit, per_page)
+        )
+        return self._make_rest_call(
+            f"{endpoint_url}{cursor_q}", action_result, method=self._req_method
+        )
+
+    def _check_datapath(self, datadict):
+        for i in list(datadict.keys()):
```
Comment (bb-splunk) [inline censys_search.py]:
```suggestion
        for i in datadict:
```

File: censys_validation.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+def parse_http_error(self, action_result, r):
+    if "json" not in r.headers.get("Content-Type", ""):
+        return ""
+
+    try:
+        resp_json = r.json()
+    except Exception:
+        return (
+            action_result.set_status(
+                phantom.APP_ERROR, "Unable to parse response as JSON"
```
Comment (bb-splunk) [inline censys_validation.py]: Why not print the response as well, whatever it was, in this case?

File: censys_validation.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    """
+    Function that checks given address and return True if address is valid IPv4 or IPV6 address.
+
+    :param input_ip_address: IP address
+    :return: status (success/failure)
+    """
+
+    try:
+        ipaddress.ip_address(input_ip_address)
+    except Exception:
```
Comment (bb-splunk) [inline censys_validation.py]:
```suggestion
    except ValueError:
```

File: censys_connector.py [hunk 4]
```diff
@@ -198,99 +95,8 @@ def _test_connectivity(self, param):
...
-            else:
-                for result in results:
-                    action_result.add_data(self._check_datapath(result))
-
-        if phantom.is_fail(ret_val):
-            return action_result.get_status(), None
-
-        return (action_result.set_status(phantom.APP_SUCCESS), response)
-
     def _handle_view(self, query_string, search_action, action_result):
```
Comment (bb-splunk) [inline censys_connector.py]: Maybe this name should also change if the API changed?

File: censys_consts.py [hunk 5]
```diff
@@ -15,11 +15,11 @@
 CENSYS_API_URL = "https://search.censys.io"
 CENSYS_API_METHOD_MAP = {
```
Comment (bb-splunk) [inline censys_consts.py:16]: Using `get` on this `dict` is rather pointless when using literals, and some of the K/V pairs appear completely unused, am I missing something

File: censys_connector.py [hunk 6]
```diff
@@ -14,127 +14,24 @@
...
 from censys_consts import *
+from censys_search import CensysSearch
+from censys_validation import get_error_message_from_exception, is_ip, parse_http_error
 
 
 class CensysConnector(BaseConnector):
     def __init__(self):
         self._headers = {}
+        self.search = CensysSearch(self._make_rest_call)
         super().__init__()
```
Comment (bbielinski-splunk) [inline censys_connector.py]: I am used to enabling the dependency injection when possible, so maybe we can consider it here too:
```suggestion
    def __init__(self, censys_search: CensysSearch | None = None):
        self._headers = {}
        self.search = censys_search if censys_search else CensysSearch(self._make_rest_call)
        super().__init__()
```

Just the suggestion tho.

File: censys_validation.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                error_msg = e.args[0]
+    except Exception:
+        self.debug_print("Error occurred while fetching exception information")
+
+    if not error_code:
+        error_text = f"Error Message: {error_msg}"
+    else:
+        error_text = f"Error Code: {error_code}. Error Message: {error_msg}"
+
+    return error_text
```
Comment (bb-splunk) [inline censys_validation.py]:
```suggestion
    return f"Error Code: {error_code}. Error Message: {error_msg}" if error_code else f"Error Message: {error_msg}"
```

File: censys_validation.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def validate_integer(action_result, parameter, key, allow_zero=False):
```
Comment (bb-splunk) [inline censys_validation.py]: Looks like the last argument is always `False` so this part of logic can be simplified. I would also move only `param = float(param) to the try/except and cover the rest with conditionals

File: censys_validation.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+# either express or implied. See the License for the specific language governing permissions
+# and limitations under the License.
+import ipaddress
+
+import phantom.app as phantom
+
+from censys_consts import CENSYS_ERR_MSG_UNAVAILABLE, CENSYS_INT_ERR_MSG
+
+
+def parse_http_error(self, action_result, r):
```
Comment (bb-splunk) [inline censys_validation.py]: No more `self` here

File: censys_validation.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    message = "Server returned error with status: {}, Type: {}, Detail: {}".format(
+        resp_json.get("status", "NA"),
+        resp_json.get("error_type", "NA"),
+        resp_json.get("error", "NA"),
+    )
+
+    return action_result.set_status(phantom.APP_ERROR, message)
+
+
+def get_error_message_from_exception(self, e):
```
Comment (bb-splunk) [inline censys_validation.py]: As above, how did this even work?

File: censys_connector.py [hunk 11]
```diff
@@ -154,14 +51,14 @@ def _make_rest_call(self, endpoint, action_result, data=None, method="post"):
...
-                        self._get_error_message_from_exception(e)
+                        get_error_message_from_exception(e)
                     ),
                 ),
                 resp_json,
             )
 
         if response.status_code not in (200, 429):
-            return (self._parse_http_error(action_result, response), {})
+            return (parse_http_error(action_result, response), {})
```
Comment (bb-splunk) [inline censys_connector.py]: Redundant parens, same below

**Review by bb-splunk — CHANGES_REQUESTED** [review]

File: censys_search.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+# either express or implied. See the License for the specific language governing permissions
+# and limitations under the License.
+import phantom.app as phantom
+
+from censys_consts import CENSYS_API_METHOD_MAP, CENSYS_JSON_QUERY, CENSYS_LIMIT_KEY
+from censys_validation import validate_integer
+
+
+class CensysSearch:
+    def __init__(self, make_rest_call) -> None:
```
Comment (bbielinski-splunk) [inline censys_search.py]: I am not a fan of passing callables to classes. Maybe we can consider pulling the i`make_rest_call` out of the connector and creating an independent function from it? As far as I see, it only requires `self.get_config()` from the connector so it should be rather easy to pass it as a parameter

**Review by bbielinski-splunk — CHANGES_REQUESTED** [review]

File: censys_connector.py (hunk 6 above)
Comment (bb-splunk) [reply censys_connector.py]: IMHO an overkill, chances of us needing more than one "censys_search" kind of object are very low

Comment (bbielinski-splunk) [reply censys_connector.py]: Maybe overkill but no cost update with potential gain

Comment (bbielinski-splunk) [reply censys_connector.py]: And also it makes it possible to have for example CensysSearch for the old censys API, current API and some possible future version of API and just switch them with one env. Who knows what will happen with this connector in the future

Comment (bb-splunk) [reply censys_connector.py]: 1) YAGNI
2) The code with dependency injection is more verbose and less readable, which would be understendable if it were needed, but it's not

File: censys_search.py (hunk 12 above)
Comment (mposluszny-splunk) [reply censys_search.py]: Yeah I am not a fan of this solution too, couldn't think of anything different here. I'll move it to different module and pass config as a param, thanks

File: censys_connector.py (hunk 6 above)
Comment (mposluszny-splunk) [reply censys_connector.py]: It's a simple change to make, but as mentioned we don't need this now.
When we'll have to cover support for multiple APIs we'll use this pattern

**Review by bbielinski-splunk — APPROVED** [review]

File: release_notes/unreleased.md [hunk 13]
```diff
@@ -1 +1,7 @@
 **Unreleased**
+
+* Updated certificate actions to work with API v2
+    * Changed pagination behavior to use 
+    * Created separate module for pagination searches, censys_search.py
+* Code refactorisation, moved helper validation to censys_validation.py
```
Comment (bb-splunk) [inline release_notes/unreleased.md]: This s an internal change and IMHO should not be in a changelog as, if done correctly, is invisible for the user

**Review by bb-splunk — APPROVED** [review]

File: release_notes/unreleased.md (hunk 13 above)
Comment (mposluszny-splunk) [reply release_notes/unreleased.md]: Thanks for clarifying, I'll write them from the user perspective

**Pull request merged**